java-checks/src/main/java/org/sonar/java/checks/StringIndexOfRangesCheck.java

```diff
@@ -0,0 +1,189 @@
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.List;
+import java.util.Optional;
+import org.sonar.check.Rule;
+import org.sonar.java.checks.methods.AbstractMethodDetection;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.JavaVersion;
+import org.sonar.plugins.java.api.JavaVersionAwareVisitor;
+import org.sonar.plugins.java.api.semantic.MethodMatchers;
+import org.sonar.plugins.java.api.tree.BinaryExpressionTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+
+@Rule(key = "S6915")
+public class StringIndexOfRangesCheck extends AbstractMethodDetection implements JavaVersionAwareVisitor {
+
+  private static final String JAVA_LANG_STRING = "java.lang.String";
+
+  private static final MethodMatchers INDEX_OF_MATCHERS =
+    MethodMatchers.create()
+      .ofTypes(JAVA_LANG_STRING)
+      .names("indexOf")
+      .addParametersMatcher("int", "int", "int")
+      .addParametersMatcher(JAVA_LANG_STRING, "int", "int")
+      .build();
+
+  private static final MethodMatchers LENGTH_MATCHERS =
+    MethodMatchers.create()
+      .ofTypes(JAVA_LANG_STRING)
+      .names("length")
+      .addWithoutParametersMatcher()
+      .build();
+
+  @Override
+  public boolean isCompatibleWithJavaVersion(JavaVersion version) {
+    return version.isJava21Compatible();
+  }
+
+  @Override
+  protected MethodMatchers getMethodInvocationMatchers() {
+    return INDEX_OF_MATCHERS;
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree methodInvocation) {
+    var issueFound = checkConstantBounds(methodInvocation);
+    if (!issueFound && methodInvocation.methodSelect() instanceof MemberSelectExpressionTree memberSelect
+      && memberSelect.expression() instanceof IdentifierTree idTree) {
+      checkBoundsThatDependOnLength(methodInvocation, idTree.name());
+    }
+  }
+
+  /**
+   * Return value indicates whether an issue has been found or not
+   */
+  private boolean checkConstantBounds(MethodInvocationTree methodInvocation) {
+    var beginIdxExpr = methodInvocation.arguments().get(1);
+    var endIdxExpr = methodInvocation.arguments().get(2);
+
+    Optional<String> receiverConst = methodInvocation.methodSelect() instanceof MemberSelectExpressionTree memberSelect
```

Comment:
We can move this line just before we actually make use of the receiverConst, to avoid useless computation